This notebook follows a deliberate 2-stage curriculum designed to reach ~75% top-1 accuracy on CIFAR-100 from scratch (no pretrained weights).

| Stage | Model | Key technique |
|-------|-------|---------------|
| 1 | ResNet-18 + augmentation + scheduler  | Baseline sanity check |
| 2 | ResNet-50 + CutMix + label smoothing | Stronger regularisation |

**Hardware:** NVIDIA T4 (Kaggle/Colab)  
**Framework:** PyTorch 2.x  
**Dataset:** CIFAR-100 — 50,000 train / 10,000 test, 100 classes, 32×32 RGB

## 0 — Imports and device

In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
import torchvision
import torchvision.transforms as transforms
import torchvision.models as models

import numpy as np
import time
import os
from tqdm import tqdm

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print('Device:', device)
if device.type == 'cuda':
    print('GPU   :', torch.cuda.get_device_name(0))
    print('VRAM  :', round(torch.cuda.get_device_properties(0).total_memory / 1e9, 1), 'GB')

Device: cuda
GPU   : Tesla T4
VRAM  : 15.6 GB


## 1 — Data loading

Three transform pipelines are defined here and reused across all stages.

- `transform_basic` — the bare minimum (used nowhere in this notebook, shown for reference)
- `transform_train` — what every stage uses for training: normalization + random crop + flip + RandAugment
- `transform_test` — deterministic: normalize only

The CIFAR-100 channel means and stds are dataset-specific constants, not ImageNet values.

In [ ]:
# CIFAR-100 channel statistics (computed over training set)
CIFAR100_MEAN = (0.5071, 0.4867, 0.4408)
CIFAR100_STD  = (0.2675, 0.2565, 0.2761)

# Training transform
# RandomCrop with padding=4: virtually enlarges each image before cropping,
# giving the model slightly shifted views of each sample.
# RandAugment: randomly applies 2 augmentation operations (rotation, shear,
# colour jitter, posterize, etc.) at magnitude 9 — the standard ViT/ResNet
# recipe for CIFAR-100.
transform_train = transforms.Compose([
    transforms.RandomCrop(32, padding=4),
    transforms.RandomHorizontalFlip(),
    transforms.RandAugment(num_ops=2, magnitude=9),
    transforms.ToTensor(),
    transforms.Normalize(CIFAR100_MEAN, CIFAR100_STD),
])

# Test transform
# No augmentation at test time — deterministic evaluation.
transform_test = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize(CIFAR100_MEAN, CIFAR100_STD),
])

# Datasets
trainset = torchvision.datasets.CIFAR100(
    root='./data', train=True,  download=True, transform=transform_train
)
testset = torchvision.datasets.CIFAR100(
    root='./data', train=False, download=True, transform=transform_test
)

# Loaders
# num_workers=2 works on Kaggle; increase to 4 on a local machine.
# pin_memory=True speeds up CPU->GPU transfer on CUDA devices.
trainloader = torch.utils.data.DataLoader(
    trainset, batch_size=128, shuffle=True,
    num_workers=2, pin_memory=True
)
testloader = torch.utils.data.DataLoader(
    testset, batch_size=256, shuffle=False,
    num_workers=2, pin_memory=True
)

print(f'Train batches : {len(trainloader)}  ({len(trainset)} samples)')
print(f'Test  batches : {len(testloader)}  ({len(testset)} samples)')

  0%|          | 0.00/169M [00:00<?, ?B/s]

  0%|          | 65.5k/169M [00:00<04:46, 589kB/s]

  0%|          | 229k/169M [00:00<02:29, 1.13MB/s]

  1%|          | 918k/169M [00:00<00:48, 3.47MB/s]

  2%|▏         | 2.65M/169M [00:00<00:20, 8.30MB/s]

  3%|▎         | 4.39M/169M [00:00<00:14, 11.3MB/s]

  4%|▎         | 6.16M/169M [00:00<00:12, 13.3MB/s]

  5%|▍         | 8.19M/169M [00:00<00:10, 15.5MB/s]

  6%|▌         | 10.4M/169M [00:00<00:09, 17.6MB/s]

  8%|▊         | 12.8M/169M [00:00<00:08, 19.4MB/s]

  9%|▉         | 15.4M/169M [00:01<00:07, 21.3MB/s]

 11%|█         | 18.1M/169M [00:01<00:06, 23.0MB/s]

 12%|█▏        | 21.0M/169M [00:01<00:05, 24.8MB/s]

 14%|█▍        | 24.1M/169M [00:01<00:05, 26.5MB/s]

 16%|█▋        | 27.6M/169M [00:01<00:04, 29.0MB/s]

 18%|█▊        | 31.3M/169M [00:01<00:04, 31.1MB/s]

 20%|██        | 34.4M/169M [00:01<00:04, 30.6MB/s]

 22%|██▏       | 37.6M/169M [00:01<00:04, 30.9MB/s]

 25%|██▍       | 41.8M/169M [00:01<00:03, 34.2MB/s]

 28%|██▊       | 46.7M/169M [00:01<00:03, 38.4MB/s]

 31%|███       | 51.8M/169M [00:02<00:02, 42.2MB/s]

 34%|███▍      | 57.1M/169M [00:02<00:02, 44.9MB/s]

 37%|███▋      | 63.0M/169M [00:02<00:02, 49.2MB/s]

 40%|████      | 68.2M/169M [00:02<00:02, 48.9MB/s]

 44%|████▍     | 74.5M/169M [00:02<00:01, 53.1MB/s]

 47%|████▋     | 79.8M/169M [00:02<00:01, 52.0MB/s]

 51%|█████     | 85.8M/169M [00:02<00:01, 52.1MB/s]

 54%|█████▍    | 91.6M/169M [00:02<00:01, 53.9MB/s]

 57%|█████▋    | 97.0M/169M [00:02<00:01, 51.9MB/s]

 60%|██████    | 102M/169M [00:03<00:01, 50.2MB/s] 

 63%|██████▎   | 107M/169M [00:03<00:01, 46.9MB/s]

 66%|██████▋   | 112M/169M [00:03<00:01, 44.8MB/s]

 69%|██████▉   | 117M/169M [00:03<00:01, 43.5MB/s]

 72%|███████▏  | 121M/169M [00:03<00:01, 42.2MB/s]

 74%|███████▍  | 125M/169M [00:03<00:01, 41.2MB/s]

 77%|███████▋  | 129M/169M [00:03<00:00, 40.3MB/s]

 79%|███████▉  | 133M/169M [00:03<00:00, 40.0MB/s]

 81%|████████▏ | 137M/169M [00:03<00:00, 39.6MB/s]

 84%|████████▎ | 141M/169M [00:04<00:00, 39.4MB/s]

 86%|████████▌ | 145M/169M [00:04<00:00, 38.8MB/s]

 88%|████████▊ | 149M/169M [00:04<00:00, 38.5MB/s]

 91%|█████████ | 153M/169M [00:04<00:00, 37.8MB/s]

 93%|█████████▎| 157M/169M [00:04<00:00, 37.6MB/s]

 95%|█████████▌| 161M/169M [00:04<00:00, 36.6MB/s]

 97%|█████████▋| 164M/169M [00:04<00:00, 35.9MB/s]

 99%|█████████▉| 168M/169M [00:04<00:00, 36.1MB/s]

100%|██████████| 169M/169M [00:04<00:00, 35.4MB/s]

Train batches : 391  (50000 samples)
Test  batches : 40  (10000 samples)


## 2 — Shared utilities

These three functions are reused across all three stages so the training loop is identical for every model.

In [ ]:
#  CutMix
# CutMix pastes a rectangular crop from one image into another, mixing their
# labels proportionally to the area of the patch.
# This forces the model to classify from partial evidence and is one of the
# single biggest regularisation gains on CIFAR-100.
def cutmix_data(images, labels, alpha=1.0):
    lam = np.random.beta(alpha, alpha)
    B, C, H, W = images.shape
    idx = torch.randperm(B).to(images.device)

    # sample a bounding box
    cut_ratio = np.sqrt(1.0 - lam)
    cut_w = int(W * cut_ratio)
    cut_h = int(H * cut_ratio)
    cx = np.random.randint(W)
    cy = np.random.randint(H)
    x1 = np.clip(cx - cut_w // 2, 0, W)
    x2 = np.clip(cx + cut_w // 2, 0, W)
    y1 = np.clip(cy - cut_h // 2, 0, H)
    y2 = np.clip(cy + cut_h // 2, 0, H)

    mixed = images.clone()
    mixed[:, :, y1:y2, x1:x2] = images[idx, :, y1:y2, x1:x2]

    # recompute lambda from actual box area
    lam = 1 - (x2 - x1) * (y2 - y1) / (W * H)
    return mixed, labels, labels[idx], lam


def cutmix_loss(criterion, outputs, labels_a, labels_b, lam):
    return lam * criterion(outputs, labels_a) + (1 - lam) * criterion(outputs, labels_b)

In [ ]:
# Universal train / evaluate functions
# use_cutmix is toggled per-stage (off for stage 1, on for stages 2 and 3)

def train_one_epoch(model, optimizer, scheduler, criterion, use_cutmix=False):
    model.train()
    total_loss = 0.0
    correct    = 0
    total      = 0

    for images, labels in trainloader:
        images, labels = images.to(device), labels.to(device)

        if use_cutmix and np.random.rand() > 0.5:
            images, la, lb, lam = cutmix_data(images, labels)
            optimizer.zero_grad()
            outputs = model(images)
            loss = cutmix_loss(criterion, outputs, la, lb, lam)
        else:
            optimizer.zero_grad()
            outputs = model(images)
            loss = criterion(outputs, labels)

        loss.backward()
        # gradient clipping — prevents occasional exploding gradients
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
        optimizer.step()
        scheduler.step()   # OneCycleLR steps every batch, not every epoch

        total_loss += loss.item()
        _, predicted = outputs.max(1)
        total   += labels.size(0)
        correct += predicted.eq(labels).sum().item()

    return total_loss / len(trainloader), 100.0 * correct / total


@torch.no_grad()
def evaluate(model):
    model.eval()
    correct = 0
    total   = 0
    for images, labels in testloader:
        images, labels = images.to(device), labels.to(device)
        outputs = model(images)
        _, predicted = outputs.max(1)
        total   += labels.size(0)
        correct += predicted.eq(labels).sum().item()
    return 100.0 * correct / total


def run_training(model, optimizer, scheduler, criterion,
                 epochs, save_path, use_cutmix=False, patience=15):
    best_acc = 0.0
    patience_counter = 0

    for epoch in range(1, epochs + 1):
        t0 = time.time()
        train_loss, train_acc = train_one_epoch(
            model, optimizer, scheduler, criterion, use_cutmix
        )
        test_acc = evaluate(model)
        elapsed  = time.time() - t0

        print(f'Epoch {epoch:3d}/{epochs} | '
              f'loss {train_loss:.4f} | '
              f'train {train_acc:.2f}% | '
              f'test {test_acc:.2f}% | '
              f'{elapsed:.0f}s')

        if test_acc > best_acc:
            best_acc = test_acc
            patience_counter = 0
            torch.save(model.state_dict(), save_path)
            print(f'  --> saved  (best {best_acc:.2f}%)')
        else:
            patience_counter += 1

        if patience_counter >= patience:
            print(f'Early stop at epoch {epoch}. Best: {best_acc:.2f}%')
            break

    print(f'\nFinal best test accuracy: {best_acc:.2f}%')
    return best_acc

## Stage 1 — ResNet-18 from scratch

1. Adding proper normalization and augmentation to the data pipeline
2. Using a cosine learning rate schedule with linear warmup (OneCycleLR)

Everything else is kept simple. No CutMix yet — evaluating how much augmentation + scheduler alone.

**ResNet-18 struggles on CIFAR-100 (when training from scratch):** it has 11M parameters but its first 7×7 conv with stride 2 immediately halves the 32×32 image to 16×16, losing a lot of spatial detail. This is designed for 224×224 images (we are trying to take care of this problem here).

In [ ]:
# ResNet-18 adjusted for 32×32 input
# The standard ResNet-18 was designed for ImageNet (224×224).
# Its first layer uses kernel=7, stride=2 and is followed by MaxPool(3,2),
# reducing 224->56 (4× downsample). On a 32×32 image this gives 8×8 after
# just the stem — too aggressive.
# Fix: replace the stem with kernel=3, stride=1, no max-pool.
# This is the standard CIFAR adaptation used in the original ResNet paper.

model_s1 = models.resnet18(weights=None)
model_s1.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
model_s1.maxpool = nn.Identity()          # remove the maxpool
model_s1.fc = nn.Linear(512, 100)         # 100 CIFAR classes
model_s1 = model_s1.to(device)

total_params = sum(p.numel() for p in model_s1.parameters())
print(f'ResNet-18 (CIFAR stem) — {total_params/1e6:.2f}M parameters')

ResNet-18 (CIFAR stem) — 11.22M parameters


In [ ]:
EPOCHS_S1 = 100

criterion_s1 = nn.CrossEntropyLoss(label_smoothing=0.1)

optimizer_s1 = optim.SGD(
    model_s1.parameters(),
    lr=0.1,                 # OneCycleLR will override this
    momentum=0.9,
    weight_decay=5e-4,
    nesterov=True           # Nesterov momentum converges slightly faster
)

# OneCycleLR: linearly warms up lr over 30% of training,
# then cosine-decays to near zero.
# This single scheduler is responsible for a large chunk of the accuracy gain
# over a flat learning rate.
scheduler_s1 = optim.lr_scheduler.OneCycleLR(
    optimizer_s1,
    max_lr=0.1,
    epochs=EPOCHS_S1,
    steps_per_epoch=len(trainloader),
    pct_start=0.3,
    anneal_strategy='cos',
    div_factor=25,           # initial lr = max_lr / 25
    final_div_factor=1e4     # final  lr = max_lr / 1e4
)

best_s1 = run_training(
    model_s1, optimizer_s1, scheduler_s1, criterion_s1,
    epochs=EPOCHS_S1,
    save_path='best_stage1_resnet18.pth',
    use_cutmix=False,
    patience=20
)

Epoch   1/100 | loss 4.3874 | train 5.13% | test 10.87% | 42s
  --> saved  (best 10.87%)


Epoch   2/100 | loss 4.0659 | train 9.97% | test 14.30% | 42s
  --> saved  (best 14.30%)


Epoch   3/100 | loss 3.8936 | train 13.10% | test 16.62% | 46s
  --> saved  (best 16.62%)


Epoch   4/100 | loss 3.7304 | train 16.46% | test 22.94% | 50s
  --> saved  (best 22.94%)


Epoch   5/100 | loss 3.5452 | train 20.06% | test 23.89% | 49s
  --> saved  (best 23.89%)


Epoch   6/100 | loss 3.3741 | train 24.03% | test 30.01% | 49s
  --> saved  (best 30.01%)


Epoch   7/100 | loss 3.2308 | train 27.40% | test 31.00% | 50s
  --> saved  (best 31.00%)


Epoch   8/100 | loss 3.0957 | train 30.49% | test 34.99% | 49s
  --> saved  (best 34.99%)


Epoch   9/100 | loss 2.9627 | train 33.95% | test 39.51% | 50s
  --> saved  (best 39.51%)


Epoch  10/100 | loss 2.8430 | train 37.03% | test 40.18% | 50s
  --> saved  (best 40.18%)


Epoch  11/100 | loss 2.7272 | train 40.42% | test 43.51% | 50s
  --> saved  (best 43.51%)


Epoch  12/100 | loss 2.6233 | train 43.31% | test 44.93% | 49s
  --> saved  (best 44.93%)


Epoch  13/100 | loss 2.5356 | train 45.72% | test 46.99% | 49s
  --> saved  (best 46.99%)


Epoch  14/100 | loss 2.4628 | train 47.70% | test 49.73% | 49s
  --> saved  (best 49.73%)


Epoch  15/100 | loss 2.3915 | train 49.58% | test 48.39% | 50s


Epoch  16/100 | loss 2.3274 | train 51.54% | test 48.51% | 50s


Epoch  17/100 | loss 2.2929 | train 52.54% | test 54.22% | 50s
  --> saved  (best 54.22%)


Epoch  18/100 | loss 2.2359 | train 54.32% | test 52.98% | 49s


Epoch  19/100 | loss 2.1907 | train 55.67% | test 51.02% | 50s


Epoch  20/100 | loss 2.1604 | train 56.19% | test 49.10% | 50s


Epoch  21/100 | loss 2.1233 | train 57.50% | test 53.74% | 50s


Epoch  22/100 | loss 2.0985 | train 58.26% | test 54.42% | 50s
  --> saved  (best 54.42%)


Epoch  23/100 | loss 2.0743 | train 59.22% | test 56.73% | 49s
  --> saved  (best 56.73%)


Epoch  24/100 | loss 2.0538 | train 59.63% | test 52.37% | 49s


Epoch  25/100 | loss 2.0320 | train 60.37% | test 55.62% | 49s


Epoch  26/100 | loss 2.0196 | train 60.73% | test 54.15% | 50s


Epoch  27/100 | loss 2.0142 | train 60.90% | test 56.60% | 49s


Epoch  28/100 | loss 1.9906 | train 61.57% | test 57.02% | 49s
  --> saved  (best 57.02%)


Epoch  29/100 | loss 1.9771 | train 61.95% | test 54.46% | 49s


Epoch  30/100 | loss 1.9674 | train 62.25% | test 52.97% | 49s


Epoch  31/100 | loss 1.9468 | train 62.82% | test 57.36% | 49s
  --> saved  (best 57.36%)


Epoch  32/100 | loss 1.9385 | train 63.28% | test 56.35% | 49s


Epoch  33/100 | loss 1.9325 | train 63.44% | test 58.44% | 50s
  --> saved  (best 58.44%)


Epoch  34/100 | loss 1.9212 | train 63.70% | test 59.03% | 49s
  --> saved  (best 59.03%)


Epoch  35/100 | loss 1.9080 | train 64.07% | test 59.59% | 49s
  --> saved  (best 59.59%)


Epoch  36/100 | loss 1.9022 | train 64.48% | test 58.88% | 49s


Epoch  37/100 | loss 1.8963 | train 64.50% | test 58.05% | 49s


Epoch  38/100 | loss 1.8841 | train 65.07% | test 58.43% | 49s


Epoch  39/100 | loss 1.8827 | train 64.79% | test 58.40% | 50s


Epoch  40/100 | loss 1.8747 | train 65.18% | test 61.62% | 49s
  --> saved  (best 61.62%)


Epoch  41/100 | loss 1.8599 | train 65.62% | test 53.91% | 49s


Epoch  42/100 | loss 1.8574 | train 65.73% | test 59.35% | 50s


Epoch  43/100 | loss 1.8493 | train 66.20% | test 60.91% | 49s


Epoch  44/100 | loss 1.8412 | train 66.31% | test 57.78% | 49s


Epoch  45/100 | loss 1.8363 | train 66.37% | test 62.79% | 49s
  --> saved  (best 62.79%)


Epoch  46/100 | loss 1.8226 | train 66.77% | test 61.09% | 49s


Epoch  47/100 | loss 1.8170 | train 66.96% | test 63.24% | 50s
  --> saved  (best 63.24%)


Epoch  48/100 | loss 1.8153 | train 66.96% | test 59.65% | 49s


Epoch  49/100 | loss 1.8040 | train 67.30% | test 61.50% | 49s


Epoch  50/100 | loss 1.7917 | train 67.83% | test 61.68% | 50s


Epoch  51/100 | loss 1.7941 | train 67.84% | test 57.36% | 49s


Epoch  52/100 | loss 1.7852 | train 67.98% | test 62.19% | 49s


Epoch  53/100 | loss 1.7757 | train 68.16% | test 59.29% | 49s


Epoch  54/100 | loss 1.7695 | train 68.48% | test 60.84% | 50s


Epoch  55/100 | loss 1.7597 | train 68.65% | test 63.33% | 50s
  --> saved  (best 63.33%)


Epoch  56/100 | loss 1.7561 | train 68.90% | test 64.59% | 49s
  --> saved  (best 64.59%)


Epoch  57/100 | loss 1.7385 | train 69.41% | test 63.03% | 49s


Epoch  58/100 | loss 1.7289 | train 69.76% | test 61.14% | 49s


Epoch  59/100 | loss 1.7334 | train 69.77% | test 65.01% | 49s
  --> saved  (best 65.01%)


Epoch  60/100 | loss 1.7121 | train 70.16% | test 64.66% | 49s


Epoch  61/100 | loss 1.7105 | train 70.32% | test 66.00% | 49s
  --> saved  (best 66.00%)


Epoch  62/100 | loss 1.6960 | train 70.91% | test 64.69% | 49s


Epoch  63/100 | loss 1.6864 | train 70.89% | test 64.58% | 49s


Epoch  64/100 | loss 1.6716 | train 71.77% | test 65.56% | 50s


Epoch  65/100 | loss 1.6570 | train 72.26% | test 63.60% | 49s


Epoch  66/100 | loss 1.6505 | train 72.27% | test 64.12% | 49s


Epoch  67/100 | loss 1.6390 | train 72.72% | test 65.08% | 50s


Epoch  68/100 | loss 1.6227 | train 73.22% | test 63.16% | 49s


Epoch  69/100 | loss 1.6139 | train 73.48% | test 65.00% | 49s


Epoch  70/100 | loss 1.5987 | train 74.01% | test 66.63% | 49s
  --> saved  (best 66.63%)


Epoch  71/100 | loss 1.5783 | train 74.69% | test 67.33% | 49s
  --> saved  (best 67.33%)


Epoch  72/100 | loss 1.5696 | train 75.00% | test 67.72% | 49s
  --> saved  (best 67.72%)


Epoch  73/100 | loss 1.5461 | train 75.85% | test 67.95% | 49s
  --> saved  (best 67.95%)


Epoch  74/100 | loss 1.5237 | train 76.47% | test 66.81% | 49s


Epoch  75/100 | loss 1.5100 | train 76.92% | test 68.66% | 49s
  --> saved  (best 68.66%)


Epoch  76/100 | loss 1.4904 | train 77.47% | test 69.70% | 49s
  --> saved  (best 69.70%)


Epoch  77/100 | loss 1.4682 | train 78.32% | test 70.12% | 49s
  --> saved  (best 70.12%)


Epoch  78/100 | loss 1.4422 | train 79.32% | test 70.84% | 49s
  --> saved  (best 70.84%)


Epoch  79/100 | loss 1.4177 | train 79.95% | test 69.96% | 49s


Epoch  80/100 | loss 1.3985 | train 80.66% | test 71.43% | 49s
  --> saved  (best 71.43%)


Epoch  81/100 | loss 1.3719 | train 81.59% | test 72.74% | 49s
  --> saved  (best 72.74%)


Epoch  82/100 | loss 1.3472 | train 82.49% | test 72.31% | 49s


Epoch  83/100 | loss 1.3147 | train 83.54% | test 73.34% | 49s
  --> saved  (best 73.34%)


Epoch  84/100 | loss 1.2879 | train 84.47% | test 73.60% | 49s
  --> saved  (best 73.60%)


Epoch  85/100 | loss 1.2647 | train 85.32% | test 74.35% | 49s
  --> saved  (best 74.35%)


Epoch  86/100 | loss 1.2278 | train 86.68% | test 75.12% | 49s
  --> saved  (best 75.12%)


Epoch  87/100 | loss 1.1956 | train 87.70% | test 74.32% | 49s


Epoch  88/100 | loss 1.1700 | train 88.60% | test 75.67% | 49s
  --> saved  (best 75.67%)


Epoch  89/100 | loss 1.1387 | train 89.83% | test 75.66% | 49s


Epoch  90/100 | loss 1.1042 | train 91.07% | test 76.75% | 49s
  --> saved  (best 76.75%)


Epoch  91/100 | loss 1.0827 | train 91.67% | test 76.59% | 49s


Epoch  92/100 | loss 1.0540 | train 92.72% | test 77.16% | 49s
  --> saved  (best 77.16%)


Epoch  93/100 | loss 1.0363 | train 93.40% | test 77.35% | 49s
  --> saved  (best 77.35%)


Epoch  94/100 | loss 1.0088 | train 94.35% | test 77.89% | 49s
  --> saved  (best 77.89%)


Epoch  95/100 | loss 0.9965 | train 94.61% | test 77.72% | 49s


Epoch  96/100 | loss 0.9791 | train 95.35% | test 78.27% | 49s
  --> saved  (best 78.27%)


Epoch  97/100 | loss 0.9707 | train 95.67% | test 78.24% | 49s


Epoch  98/100 | loss 0.9621 | train 95.97% | test 78.41% | 49s
  --> saved  (best 78.41%)


Epoch  99/100 | loss 0.9606 | train 96.00% | test 78.28% | 49s


Epoch 100/100 | loss 0.9571 | train 96.15% | test 78.22% | 49s

Final best test accuracy: 78.41%


## Stage 2 — ResNet-50 + CutMix + label smoothing

Two changes from Stage 1:
1. **Larger backbone** — ResNet-50 (25M params vs 11M) with the same CIFAR stem fix
2. **CutMix** — applied randomly to 50% of batches. This is the single most impactful regularisation technique for CIFAR-100 and prevents the model from over-relying on small discriminative regions

Label smoothing (0.1) is carried over from Stage 1. It prevents overconfidence by replacing hard 0/1 targets with 0.001/0.901, which reduces overfitting on a 100-class problem.

In [ ]:
# ResNet-50 adjusted for 32×32 input
model_s2 = models.resnet50(weights=None)
model_s2.conv1   = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
model_s2.maxpool = nn.Identity()
model_s2.fc      = nn.Linear(2048, 100)
model_s2 = model_s2.to(device)

total_params = sum(p.numel() for p in model_s2.parameters())
print(f'ResNet-50 (CIFAR stem) — {total_params/1e6:.2f}M parameters')

ResNet-50 (CIFAR stem) — 23.71M parameters


In [ ]:
EPOCHS_S2 = 120

criterion_s2 = nn.CrossEntropyLoss(label_smoothing=0.1)

# SGD + momentum is the standard for ResNets on CIFAR.
# Weight decay slightly higher than Stage 1 to counter the larger model.
optimizer_s2 = optim.SGD(
    model_s2.parameters(),
    lr=0.1,
    momentum=0.9,
    weight_decay=1e-3,
    nesterov=True
)

scheduler_s2 = optim.lr_scheduler.OneCycleLR(
    optimizer_s2,
    max_lr=0.1,
    epochs=EPOCHS_S2,
    steps_per_epoch=len(trainloader),
    pct_start=0.3,
    anneal_strategy='cos',
    div_factor=25,
    final_div_factor=1e4
)

best_s2 = run_training(
    model_s2, optimizer_s2, scheduler_s2, criterion_s2,
    epochs=EPOCHS_S2,
    save_path='best_stage2_resnet50.pth',
    use_cutmix=True,   # CutMix enabled
    patience=20
)

Epoch   1/120 | loss 4.6172 | train 1.31% | test 1.63% | 195s
  --> saved  (best 1.63%)


Epoch   2/120 | loss 4.5979 | train 1.47% | test 1.86% | 193s


  --> saved  (best 1.86%)


Epoch   3/120 | loss 4.5802 | train 1.85% | test 2.74% | 194s


  --> saved  (best 2.74%)


Epoch   4/120 | loss 4.5402 | train 2.50% | test 4.39% | 194s


  --> saved  (best 4.39%)


Epoch   5/120 | loss 4.4523 | train 3.71% | test 6.60% | 193s


  --> saved  (best 6.60%)


Epoch   6/120 | loss 4.3185 | train 5.92% | test 10.95% | 194s


  --> saved  (best 10.95%)


Epoch   7/120 | loss 4.2362 | train 7.24% | test 12.35% | 193s


  --> saved  (best 12.35%)


Epoch   8/120 | loss 4.1312 | train 9.38% | test 14.93% | 194s


  --> saved  (best 14.93%)


Epoch   9/120 | loss 4.0062 | train 11.72% | test 18.81% | 194s


  --> saved  (best 18.81%)


Epoch  10/120 | loss 3.8789 | train 14.60% | test 22.49% | 194s


  --> saved  (best 22.49%)


Epoch  11/120 | loss 3.7403 | train 17.27% | test 27.11% | 193s


  --> saved  (best 27.11%)


Epoch  12/120 | loss 3.5729 | train 21.04% | test 31.80% | 194s


  --> saved  (best 31.80%)


Epoch  13/120 | loss 3.4279 | train 24.66% | test 33.55% | 194s


  --> saved  (best 33.55%)


Epoch  14/120 | loss 3.3083 | train 28.15% | test 35.80% | 193s


  --> saved  (best 35.80%)


Epoch  15/120 | loss 3.2025 | train 30.59% | test 38.26% | 193s


  --> saved  (best 38.26%)


Epoch  16/120 | loss 3.1124 | train 33.44% | test 38.32% | 193s


  --> saved  (best 38.32%)


Epoch  17/120 | loss 3.0791 | train 34.32% | test 41.18% | 194s


  --> saved  (best 41.18%)


Epoch  18/120 | loss 3.0083 | train 35.52% | test 34.86% | 194s


Epoch  19/120 | loss 2.9707 | train 37.21% | test 42.09% | 194s
  --> saved  (best 42.09%)


Epoch  20/120 | loss 2.9866 | train 36.66% | test 41.50% | 193s


Epoch  21/120 | loss 2.9888 | train 36.97% | test 38.37% | 193s


Epoch  22/120 | loss 2.9625 | train 37.77% | test 45.01% | 193s
  --> saved  (best 45.01%)


Epoch  23/120 | loss 2.9811 | train 37.45% | test 42.31% | 194s


Epoch  24/120 | loss 2.9271 | train 38.86% | test 44.02% | 193s


Epoch  25/120 | loss 2.9616 | train 37.83% | test 46.33% | 193s
  --> saved  (best 46.33%)


Epoch  26/120 | loss 2.8914 | train 39.87% | test 40.08% | 193s


Epoch  27/120 | loss 2.9441 | train 38.02% | test 36.98% | 193s


Epoch  28/120 | loss 2.8987 | train 38.80% | test 39.38% | 193s


Epoch  29/120 | loss 2.9284 | train 39.04% | test 40.19% | 193s


Epoch  30/120 | loss 2.8984 | train 40.39% | test 39.81% | 192s


Epoch  31/120 | loss 2.9324 | train 39.07% | test 43.31% | 193s


Epoch  32/120 | loss 2.8890 | train 39.87% | test 42.40% | 193s


Epoch  33/120 | loss 2.8625 | train 41.14% | test 41.99% | 193s


Epoch  34/120 | loss 2.8817 | train 40.04% | test 41.18% | 193s


Epoch  35/120 | loss 2.8225 | train 41.90% | test 38.35% | 193s


Epoch  36/120 | loss 2.8586 | train 40.80% | test 44.01% | 193s


Epoch  37/120 | loss 2.8102 | train 41.99% | test 43.42% | 193s


Epoch  38/120 | loss 2.8626 | train 40.30% | test 43.52% | 193s


Epoch  39/120 | loss 2.8733 | train 40.49% | test 46.72% | 192s


  --> saved  (best 46.72%)


Epoch  40/120 | loss 2.8519 | train 40.81% | test 40.42% | 192s


Epoch  41/120 | loss 2.8240 | train 42.21% | test 42.32% | 193s


Epoch  42/120 | loss 2.8495 | train 41.21% | test 47.97% | 193s


  --> saved  (best 47.97%)


Epoch  43/120 | loss 2.7880 | train 42.91% | test 44.24% | 192s


Epoch  44/120 | loss 2.8700 | train 40.67% | test 48.03% | 192s


  --> saved  (best 48.03%)


Epoch  45/120 | loss 2.8545 | train 42.02% | test 44.53% | 192s


Epoch  46/120 | loss 2.8173 | train 41.93% | test 42.50% | 192s


Epoch  47/120 | loss 2.7710 | train 42.85% | test 47.86% | 193s


Epoch  48/120 | loss 2.7971 | train 42.45% | test 47.85% | 192s


Epoch  49/120 | loss 2.7797 | train 42.92% | test 45.85% | 192s


Epoch  50/120 | loss 2.7979 | train 42.27% | test 47.33% | 192s


Epoch  51/120 | loss 2.7727 | train 43.77% | test 41.95% | 192s


Epoch  52/120 | loss 2.8126 | train 41.72% | test 47.63% | 192s


Epoch  53/120 | loss 2.7674 | train 43.06% | test 47.53% | 192s


Epoch  54/120 | loss 2.7654 | train 43.78% | test 45.52% | 193s


Epoch  55/120 | loss 2.7827 | train 43.00% | test 41.41% | 193s


Epoch  56/120 | loss 2.7623 | train 43.71% | test 48.80% | 193s


  --> saved  (best 48.80%)


Epoch  57/120 | loss 2.7636 | train 43.34% | test 46.63% | 192s


Epoch  58/120 | loss 2.7199 | train 44.45% | test 47.43% | 193s


Epoch  59/120 | loss 2.7592 | train 43.44% | test 51.57% | 193s
  --> saved  (best 51.57%)


Epoch  60/120 | loss 2.7243 | train 44.45% | test 48.42% | 193s


Epoch  61/120 | loss 2.7729 | train 43.58% | test 49.80% | 193s


Epoch  62/120 | loss 2.7837 | train 43.12% | test 48.22% | 193s


Epoch  63/120 | loss 2.7663 | train 43.11% | test 48.36% | 193s


Epoch  64/120 | loss 2.6999 | train 45.61% | test 51.97% | 193s


  --> saved  (best 51.97%)


Epoch  65/120 | loss 2.7214 | train 45.10% | test 53.56% | 193s


  --> saved  (best 53.56%)


Epoch  66/120 | loss 2.6936 | train 44.84% | test 49.93% | 193s


Epoch  67/120 | loss 2.7419 | train 44.63% | test 45.12% | 193s


Epoch  68/120 | loss 2.7074 | train 45.53% | test 50.78% | 193s


Epoch  69/120 | loss 2.6610 | train 47.13% | test 48.49% | 193s


Epoch  70/120 | loss 2.6849 | train 45.75% | test 52.35% | 193s


Epoch  71/120 | loss 2.7027 | train 45.70% | test 54.35% | 193s


  --> saved  (best 54.35%)


Epoch  72/120 | loss 2.7114 | train 45.47% | test 53.30% | 193s


Epoch  73/120 | loss 2.6540 | train 47.02% | test 52.80% | 193s


Epoch  74/120 | loss 2.6737 | train 46.37% | test 52.73% | 193s


Epoch  75/120 | loss 2.6312 | train 47.04% | test 56.99% | 193s
  --> saved  (best 56.99%)


Epoch  76/120 | loss 2.6640 | train 46.30% | test 53.77% | 193s


Epoch  77/120 | loss 2.6941 | train 46.01% | test 56.63% | 193s


Epoch  78/120 | loss 2.6823 | train 46.06% | test 56.05% | 193s


Epoch  79/120 | loss 2.6442 | train 47.76% | test 53.84% | 193s


Epoch  80/120 | loss 2.6275 | train 48.34% | test 53.65% | 193s


Epoch  81/120 | loss 2.6016 | train 48.69% | test 53.57% | 193s


Epoch  82/120 | loss 2.6187 | train 48.72% | test 57.22% | 193s


  --> saved  (best 57.22%)


Epoch  83/120 | loss 2.6073 | train 48.91% | test 59.00% | 193s


  --> saved  (best 59.00%)


Epoch  84/120 | loss 2.5795 | train 49.36% | test 58.65% | 193s


Epoch  85/120 | loss 2.5819 | train 50.13% | test 58.38% | 193s


Epoch  86/120 | loss 2.5472 | train 51.08% | test 60.67% | 193s


  --> saved  (best 60.67%)


Epoch  87/120 | loss 2.5722 | train 49.29% | test 59.49% | 193s


Epoch  88/120 | loss 2.5424 | train 50.89% | test 58.95% | 193s


Epoch  89/120 | loss 2.5044 | train 52.34% | test 61.40% | 193s
  --> saved  (best 61.40%)


Epoch  90/120 | loss 2.5209 | train 51.64% | test 58.70% | 193s


Epoch  91/120 | loss 2.5181 | train 52.20% | test 62.50% | 193s
  --> saved  (best 62.50%)


Epoch  92/120 | loss 2.4735 | train 52.93% | test 62.76% | 193s


  --> saved  (best 62.76%)


Epoch  93/120 | loss 2.4439 | train 53.22% | test 60.35% | 193s


Epoch  94/120 | loss 2.3927 | train 54.91% | test 65.61% | 193s


  --> saved  (best 65.61%)


Epoch  95/120 | loss 2.4925 | train 51.68% | test 63.98% | 193s


Epoch  96/120 | loss 2.3710 | train 55.25% | test 65.70% | 193s
  --> saved  (best 65.70%)


Epoch  97/120 | loss 2.3550 | train 56.51% | test 65.95% | 193s


  --> saved  (best 65.95%)


Epoch  98/120 | loss 2.4075 | train 55.30% | test 64.70% | 193s


Epoch  99/120 | loss 2.3387 | train 57.62% | test 68.77% | 193s
  --> saved  (best 68.77%)


Epoch 100/120 | loss 2.2806 | train 58.00% | test 68.39% | 193s


Epoch 101/120 | loss 2.2696 | train 59.67% | test 67.82% | 193s


Epoch 102/120 | loss 2.2359 | train 60.62% | test 70.92% | 193s
  --> saved  (best 70.92%)


Epoch 103/120 | loss 2.2273 | train 60.80% | test 71.35% | 193s
  --> saved  (best 71.35%)


Epoch 104/120 | loss 2.1674 | train 62.83% | test 71.26% | 193s


Epoch 105/120 | loss 2.1928 | train 62.98% | test 72.28% | 193s


  --> saved  (best 72.28%)


Epoch 106/120 | loss 2.2070 | train 63.20% | test 73.81% | 193s


  --> saved  (best 73.81%)


Epoch 107/120 | loss 2.1271 | train 64.22% | test 74.56% | 193s
  --> saved  (best 74.56%)


Epoch 108/120 | loss 2.0529 | train 67.00% | test 76.37% | 193s
  --> saved  (best 76.37%)


Epoch 109/120 | loss 2.0311 | train 67.85% | test 75.98% | 193s


Epoch 110/120 | loss 2.0600 | train 66.75% | test 77.41% | 193s


  --> saved  (best 77.41%)


Epoch 111/120 | loss 1.9283 | train 69.92% | test 78.14% | 193s
  --> saved  (best 78.14%)


Epoch 112/120 | loss 1.9343 | train 72.22% | test 77.86% | 193s


Epoch 113/120 | loss 1.9207 | train 71.77% | test 79.72% | 193s
  --> saved  (best 79.72%)


Epoch 114/120 | loss 1.9173 | train 71.88% | test 80.05% | 193s
  --> saved  (best 80.05%)


Epoch 115/120 | loss 1.8794 | train 72.78% | test 80.67% | 193s


  --> saved  (best 80.67%)


Epoch 116/120 | loss 1.8814 | train 72.90% | test 81.01% | 193s


  --> saved  (best 81.01%)


Epoch 117/120 | loss 1.8168 | train 74.94% | test 81.10% | 193s
  --> saved  (best 81.10%)


Epoch 118/120 | loss 1.8496 | train 74.26% | test 81.32% | 193s
  --> saved  (best 81.32%)


Epoch 119/120 | loss 1.8238 | train 75.01% | test 81.49% | 193s
  --> saved  (best 81.49%)


Epoch 120/120 | loss 1.8136 | train 76.33% | test 81.46% | 193s

Final best test accuracy: 81.49%


## 3 Results Summary

| Model | Params | Train acc | Test acc | Gap | Status |
|-------|--------|-----------|----------|-----|--------|
| ResNet-18 | 11M | 96.15% | 78.41% | +17.74% | Capacity-limited |
| ResNet-50 + CutMix | 25M | 76.33% | 81.49% | -5.16% | Still converging |

**Key takeaway:** ResNet-50 with CutMix is the clear winner and the model to carry forward.
The inverted gap on ResNet-50 confirms CutMix is working as intended. ResNet-18 is retired —
its overfitting gap is structural and cannot be resolved without architectural changes.
